draft-irtf-cfrg-vdaf.md

```diff
@@ -2275,281 +2281,654 @@ the protocol works as follows.
+concrete instantiation of Poplar1 is specified in {{poplar1-instantiation}};
+test vectors can be found in {{test-vectors}}.
+
+## Incremental Distributed Point Functions (IDPFs) {#idpf}
 
 An IDPF is defined over a domain of size `2^BITS`, where `BITS` is constant
-defined by the IDPF. The Client specifies an index `alpha` and a vector of values
-`beta`, one for each "level" `1 <= L <= BITS`. The key generation algorithm generates two
-IDPF keys, one for each Aggregator. When evaluated at level `L` and index `0 <= x < 2^L`, each
-IDPF share returns an additive share of `beta[L]` if `x` is the `L`-bit prefix
-of `alpha` and shares of zero otherwise.
-
-Let `LSB(x, N)` denote the least significant `N` bits of positive integer
-`x`. By definition, a positive integer `0 <= x <= 2^L` is said to be the length-`L`
+defined by the IDPF. Indexes into the IDPF tree are encoded as integers in range
+`[0, 2^BITS)`. The Client specifies an index `alpha` and a vector of
+values `beta`, one for each "level" `L` in range `[0, BITS)`. The key generation
+algorithm generates one IDPF "key" for each Aggregator. When evaluated at level
+`L` and index `0 <= prefix < 2^L`, each IDPF key returns an additive share of
+`beta[L]` if `prefix` is the `L`-bit prefix of `alpha` and shares of zero
+otherwise.
+
+An index `x` is defined to be a prefix of another index `y` as follows. Let
+`LSB(x, N)` denote the least significant `N` bits of positive integer `x`. By
+definition, a positive integer `0 <= x <= 2^L` is said to be the length-`L`
 prefix of positive integer `0 <= y <= BITS` if `LSB(x, L)` is equal to the most
-significant `L` bits of `LSB(y, BITS)`, For example, 6 (110 in binary) is
-the length-3 prefix of 25 (11001), but 7 (111) is not.
+significant `L` bits of `LSB(y, BITS)`, For example, 6 (110 in binary) is the
+length-3 prefix of 25 (11001), but 7 (111) is not.
+
+Each of the programmed points `beta` is a vector of elements of some finite
+field. We distinguish two types of fields: One for inner nodes (denoted
+`Idpf.FieldInner`), and one for leaf nodes (`Idpf.FieldLeaf`). (Our
+instantiation of Poplar1 ({{poplar1-instantiation}}) will use a much larger
+field for leaf nodes than for inner nodes. This is to ensure the IDPF is
+"extractable" as defined in {{BBCGGI21}}, Definition 1.)
 
-Each `beta[l]` is a vector of elements of a finite field. We distinguish two types of
-fields: one for inner nodes (denoted `Idpf.FieldInner`), and one for leaf nodes (`Idpf.FieldLeaf`).
+A concrete IDPF defines the types and constants enumerated in {{idpf-param}}. In
+the remainder we write `Idpf.Vec` as shorthand for the type
+`Union[Vec[Vec[Idpf.FieldInner]], Vec[Vec[Idpf.FieldLeaf]]]`. (This type denotes
+either a vector of inner node field elements or leaf node field elements.) The
+scheme is comprised of the following algorithms:
 
-An IDPF is comprised of the following algorithms :
+* `Idpf.gen(alpha: Unsigned, beta_inner: Vec[Vec[Idpf.FieldInner]], beta_leaf:
```

Comment:
`alpha` is the index being sharded here, right?  If so, it would be clearer to call it a bit string, not an integer, since that's how it is referred to above.  (Maybe internally to the function it is converted to an integer, but that doesn't really matter as long as all participants put the bits in the same order.)

Where do the `beta` values come from?  If they're computed from `alpha`, it seems like there's a missing step for that.  If they're constant / defined by the IDPF, then it seems like they can be treated as internal to the function, not arguments.

Reply:
> `alpha` is the index being sharded here, right? If so, it would be clearer to call it a bit string, not an integer, since that's how it is referred to above. (Maybe internally to the function it is converted to an integer, but that doesn't really matter as long as all participants put the bits in the same order.)

@schoppmp and I discussed this a bit and decided that `Unsigned` was the best representation of the input. A couple of alternatives:
* `Bytes` - This is the most intuitive, however it has the problem of not being able to represent a bit strings whose length is not a multiple of 8. Of course we could consider requiring input strings be a multiple 8. This would probably be fine for most applications.
* Define some sort of `BitString` type, perhaps even making use of ASN.1. We decided this would be a bit too complicated.
 
> Where do the `beta` values come from? If they're computed from `alpha`, it seems like there's a missing step for that. If they're constant / defined by the IDPF, then it seems like they can be treated as internal to the function, not arguments.

The `beta` values are arbitrary and are picked by the caller. In `Poplar1`, they are chosen in a particular way.

Reply:
On the first point: I'm not overly picky here.  Maybe you could just say somewhere that the Unsigned value is the input bit string of length `BITS`, interpreted as a `BITS`-bit big-endian integer (or whatever, doesn't really matter).

On the second point: If they're (a) generated unilaterally by the caller and (b) not used elsewhere, they should get generated internally to `gen`, not passed in as arguments.

Reply:
> On the first point: I'm not overly picky here. Maybe you could just say somewhere that the Unsigned value is the input bit string of length `BITS`, interpreted as a `BITS`-bit big-endian integer (or whatever, doesn't really matter).

Yeah we should provide some guidance here, but at this point I'm not sure what to recommend here. Really any encoding is fine, so long as the application is consistent.


> On the second point: If they're (a) generated unilaterally by the caller and (b) not used elsewhere, they should get generated internally to `gen`, not passed in as arguments.

Well, one could imagine different applications for IDPFs that call for different values to be "programmed". For Poplar1, the programmed values are pairs `(1, k)` where `k` is a random field element used for verifying the output shares. When used in a plain DAF, there would be no need for the random field element. One could also imagine an application in which "weighted" inputs are used, i.e., rather than always output `1`, output some integer in range `[0, c)` chosen by the user.

Another thing to note is that the current IDPF syntax more or less matches the paper.